File: src/dotnet/Core/Application/Services/SkillNormalizer.cs
using System.Text.RegularExpressions;

namespace Rag.Candidates.Core.Application.Services;

/// <summary>
/// Normalizes skill names to canonical technology names for deterministic filtering.
/// </summary>
public static class SkillNormalizer
{
    private static readonly Dictionary<string, string> NormalizationRules = new()
    {
        // Java variants
        [@"^Java\s*\d*\.?\d*$"] = "Java",
        [@"^Java\s*\([^)]+\)$"] = "Java",
        [@".*\bJava\b.*"] = "Java",
        
        // JavaScript variants
        [@"^JavaScript.*"] = "JavaScript",
        [@"^JS\s*\(.*\)$"] = "JavaScript",
        
        // Python variants
        [@"^Python\s*\d*\.?\d*$"] = "Python",
        
        // C# variants
        [@"^C#.*"] = "C#",
        [@"^CSharp.*"] = "C#",
        
        // TypeScript variants
        [@"^TypeScript.*"] = "TypeScript",
        [@"^TS\s*\(.*\)$"] = "TypeScript",
        
        // React variants
        [@"^React.*"] = "React",
        
        // Angular variants
        [@"^Angular.*"] = "Angular",
        
        // Spring variants
        [@"^Spring\s*\([^)]+\)$"] = "Spring",
        [@"^Spring\s+\w+$"] = "Spring",
        
        // .NET variants
        [@"^\.NET.*"] = ".NET",
        [@"^ASP\.NET.*"] = "ASP.NET",
        
        // Node variants
        [@"^Node\.?js.*"] = "Node.js",
        
        // SQL variants
        [@"^SQL\s*\([^)]+\)$"] = "SQL",
        [@".*SQL$"] = "SQL",
        
        // Docker
        [@"^Docker.*"] = "Docker",
        
        // Kubernetes
        [@"^Kubernetes.*"] = "Kubernetes",
        [@"^K8s$"] = "Kubernetes",
        
        // Git
        [@"^Git.*"] = "Git",
    };

    /// <summary>
    /// Normalize a skill name to its canonical form.
    /// </summary>
    /// <param name="skillName">Raw skill name from candidate data</param>
    /// <returns>Canonical skill name for filtering, or original if no rule matches</returns>
    public static string Normalize(string skillName)
    {
        if (string.IsNullOrWhiteSpace(skillName))
            return skillName;

        var skillTrimmed = skillName.Trim();

        foreach (var (pattern, canonical) in NormalizationRules)
        {
            if (Regex.IsMatch(skillTrimmed, pattern, RegexOptions.IgnoreCase))
            {
                return canonical;
            }
        }

        return skillTrimmed;
    }
}

File: src/python/core/application/services/skill_normalizer.py
"""Normalize skill names to canonical technology names for deterministic filtering."""
import re
from typing import Dict


class SkillNormalizer:
    """Normalizes skill names to canonical technology identifiers."""
    
    _NORMALIZATION_RULES: Dict[str, str] = {
        # Java variants
        r'^Java\s*\d*\.?\d*$': 'Java',
        r'^Java\s*\([^)]+\)$': 'Java',
        r'.*\bJava\b.*': 'Java',
        
        # JavaScript variants
        r'^JavaScript.*': 'JavaScript',
        r'^JS\s*\(.*\)$': 'JavaScript',
        
        # Python variants
        r'^Python\s*\d*\.?\d*$': 'Python',
        
        # C# variants
        r'^C#.*': 'C#',
        r'^CSharp.*': 'C#',
        
        # TypeScript variants
        r'^TypeScript.*': 'TypeScript',
        r'^TS\s*\(.*\)$': 'TypeScript',
        
        # React variants
        r'^React.*': 'React',
        
        # Angular variants
        r'^Angular.*': 'Angular',
        
        # Spring variants
        r'^Spring\s*\([^)]+\)$': 'Spring',
        r'^Spring\s+\w+$': 'Spring',
        
        # .NET variants
        r'^\.NET.*': '.NET',
        r'^ASP\.NET.*': 'ASP.NET',
        
        # Node variants
        r'^Node\.?js.*': 'Node.js',
        
        # SQL variants
        r'^SQL\s*\([^)]+\)$': 'SQL',
        r'.*SQL$': 'SQL',
        
        # Docker
        r'^Docker.*': 'Docker',
        
        # Kubernetes
        r'^Kubernetes.*': 'Kubernetes',
        r'^K8s$': 'Kubernetes',
        
        # Git
        r'^Git.*': 'Git',
    }
    
    @staticmethod
    def normalize(skill_name: str) -> str:
        """
        Normalize a skill name to its canonical form.
        
        Args:
            skill_name: Raw skill name from candidate data
            
        Returns:
            Canonical skill name for filtering, or original if no rule matches
        """
        if not skill_name:
            return skill_name
        
        skill_trimmed = skill_name.strip()
        
        for pattern, canonical in SkillNormalizer._NORMALIZATION_RULES.items():
            if re.match(pattern, skill_trimmed, re.IGNORECASE):
                return canonical
        
        return skill_trimmed
